### PR TITLE
CRM_Utils_Cache - Use more consistent group names

### PR DESCRIPTION
**Pull request by totten**
Overview
----------------------------------------

This is a batch of refinements for how `civicrm_cache` groups are named.

Before + After (Effective Cache Names)
----------------------------------------

| `old_group_name`                   | `new_group_name`                   |
|------------------------------------|-------------------------------------|
| `698ba0f63041a6f1a88d60caf82c17e7` | `-ext_QpVei68UFYfiMP1l-hLBu2Bx5`    |
| `angular_5c6c14c55af45f3ba5b78a02` | `angular/5.80a1`    |
| `CiviCRM-20Session`                | ~~`CiviCRM Session/5.80a1`~~ <br/> `CiviCRM Session`            |
| `contact-20fields`                 | `contact fields/5.80a1`             |
| `contactTypes`                     | `contactTypes/5.80a1`               |
| `js_strings`                       | `js_strings/5.80a1`                 |
| `long`                             | `long/5.80a1`                       |
| `metadata_5_80_alpha1`             | `metadata/5.80a1`                   |
| `settings_5_80_alpha1`             | `settings/5.80a1`                   |

Before + After (Coding Style)
----------------------------------------

* Before, the `settings` and `metadata` and `ext` caches each had different logic to apply a suffix.
    ```php
    $cache = CRM_Utils_Cache::create([
      'name' => 'foo_' . figureOutVersionNumber(),
    ]);
    ```
* After, they let the cache-helper figure out what to append.
    ```php
    $cache = CRM_Utils_Cache::create([
      'name' => 'foo',
      'scope' => 'version',
    ]);
    ```

Technical Details
----------------------------------------

1. When faced with a long group-name, try harder to preserve a recognizable prefix.
    * In the example, the extension-cache has a long name.
    * To fit within the SQL limits while preserving uniqueness, the name was hashed -- yielding the completely inscrutable value `698ba0f63041a6f1a88d60caf82c17e7`. 
    * Now, it's still looks funny, but it retains a prefix which hints at the content. (`ext_42955e8ba23Hvwkl1xaFlEtxA` is probably related to extensions.)
    * Additionally, it uses shorter encoding for the hash (Base64 vs Base16) to improve the amount of prefix-data that we can display.
2. Allow some more common characters in SQL group name
    * In the example, note how the name `CiviCRM-20Session` differs from `CiviCRM Session`.
    * This is because the escaping rule was tailored to PSR-16, and PSR-16 doesn't defined space (` `) for use in keys..
    * However, SQL group-names aren't really subject to PSR-16 restrictions. When setting up the `SqlGroup` object, we can allow a slightly more relaxed formula.
3. Trim the version code
    * Shorter expressions are less likely to provoke hashing. "5.80a1" and  "5.80.alpha1" are equally readable.
4.  Expand on the idea from #31399 regarding upgrades -- Several caches are (initially) re-used across different versions, but they probably shouldn't be.
    * 1-by-1, we've been adding version-code to cache-groups (settings, metadata, extensions, etc). But actually the technique makes sense for several more caches (js_strings, contact fields, afform paths, etc).
5. Include the idea of https://lab.civicrm.org/dev/core/-/issues/5436 
    * The Angular cache can now be shared between different runtimes (eg between WP frontend and WP backend).

Should we be optimistic?  I expect Jenkins to have some initial complaints, but I'm fairly optimistic.
    * Basically -- Civi already supports multiple cache-backends (SQL/Redis/Memcache). Storage is abstracted around `$cache` objects. The encoding of the group-name is an internal detail of the cache-object and shouldn't matter much for consumers of the `$cache` API.
    * There is one quirky cache (`CiviCRM Search PrevNextCache`) which I left as-is because it has funny rules.

Comments
----------------------------------------

Expanding the size of `civicrm_cache.group_name` may also be a good idea. But some limit will still exist, so it's still fine to tune these rules.

### CODE REVIEW

**Comment by eileenmcnaughton** [comment]
@totten there is one more scope concept - `CIVICRM_DEPLOY_ID` - I don't think this change alters anything regarding deploy ID - but I think it might make sense to describe it's significance in the code comments (basically to ensure that re-running civi-build doesn't connect to the previous build's redis cache)

Regarding the default - I feel like a narrower default is better than global  - generally I would go for

- Domain ID + Version ID OR
- Domain ID + version ID + locale

(just kicking the tyres mentally on the latter - but for `metadata` I think the latter makes more sense)

**Comment by eileenmcnaughton** [comment]
SO thinking about this more - I had been thinking about resolving https://github.com/civicrm/civicrm-core/pull/31409 by finding the right cache key & doing something like we do with tokens, option values, various other cache keys - ie 

```
  $cacheKey = __CLASS__ . 'domain_tokens' . $html . '_' . $domainID . '_' . CRM_Core_I18n::getLocale();
    if (!Civi::cache('metadata')->has($cacheKey)) {
```

However, I'm now thinking that per-locale cache separation like this makes more sense ... except that when we do deliberately clear the metadata cache I guess we do want to clear all the locales - so maybe it doesn't

**Comment by eileenmcnaughton** [comment]
e.g cache clear in `civi_import` (would be nice to improve) - but this is because the imports are entities

image

**Comment by ufundo** [comment]
really like this initiative!

my initial attempted test of an upgrade Standalone 5.78 => 5.79 didn't go to plan, but I think there may have been external factors. will try again later in the week.

in the meantime:

could the case / spaces be standardised at the same time?
image

 
> "5.80a1" and "5.80.alpha1" are equally readable. 

I don't agree, I think alpha is, in context, self explanatory, and what one is used to seeing elsewhere. Can we not afford the extra 4 characters ??

**Comment by eileenmcnaughton** [comment]
just noting that @totten & discussed this a bit & I feel like we landed on NOT having the global scope - or if we do only as a default and always using the version in the scope. In general there are risks of NOT including the version that I feel we understand.

@totten & I discussed whether there were risks in including the version  & the scenario Tim could see was that is someone was filling in a contribution page at the exact moment the codebase was being replaced their form session would be lost if the session cache is no longer in global scope. There are flip-side risks to their form session NOT being lost if their save becomes invalid for some reason. However, we discovered that the session cache is currently cleared every time a settings form is saved  & people have not raised this as a problem then losing sessions at a time when site owners are already recommended to put the site in maintenance mode feels like it's on the right side of the risk calculation

**Comment by totten** [comment]
This comment is about one specific cache (`CiviCRM Session`, aka "_active form-states for QuickForm_") and not the PR generally. This is a fairly unusual cache because it actually tracks *interactive state* and not just *dummy/rebuildable data*.

> ... risk ... someone was filling in a contribution page at the exact moment the codebase was being replaced  ...

Some free-association...

* When I submitted this PR, I was assuming that the upgrader was already flushing `CiviCRM Session`. And I think it did do that for a long time. (*But not anymore.*)
* The question of whether to flush form-state when upgrading is fundamentally murky. Systemically speaking, it is  more *consistent/predictable* to flush; but *in most typical cases*, it's *unnecessary* and *a little harmful*.
* The relevant code-paths/history are convoluted/whack-a-mole-y.
* I suspect ecf7b3ba6e1 flipped the upgrade-behavior circa March 2024. It sort-of created a carve-out for `CiviCRM Session`. In that discussion, *upgrades* weren't actually the focus. (But thematically, it was in the vein of "keep sessions longer.")
* If I'd recognized that it had a special carve-out at the moment when I wrote this patch, then I probably would've put a different policy on that cache.

Conclusion: So, OK, since `CiviCRM Session` (QF-form-state) is kind of special, I'll switch it back to retain state across upgrades.

**Comment by eileenmcnaughton** [comment]
@totten I'm OK to merge this now - my one reservation is that I think `version` rather than `global` would be a safer default - but I see that my previous comment suggested we agreed on `global` as a default - giving it merge-ready in case it's worth switching

**Comment by totten** [comment]
> my one reservation is that I think version rather than global would be a safer default

For `civicrm-core` or for greenfield code, I would agree. OTOH, looking at universe makes me a little nervous -- I see ~30 references to `CRM_Utils_Cache::create()` in extensions. I tried skimming these for a sense of impact, but you wind up needing to follow-up searches and thinking about more files.

Defaulting to `global` makes it a drop-in equivalent.

A policy change for contrib might not be a bad thing, but that can be another day.
